src/mpid/ch4/ch4_api.txt

```diff
@@ -0,0 +1,532 @@
+     SHM*: sendbuf, sendcount, sendtype, recvbuf, recvcount, recvtype, root, comm, req-2
+  mpi_iscatterv : int
+      NM*: sendbuf, sendcounts, displs, sendtype, recvbuf, recvcount, recvtype, root, comm_ptr, req-2
+     SHM*: sendbuf, sendcounts, displs, sendtype, recvbuf, recvcount, recvtype, root, comm_ptr, req-2
+  mpi_type_commit_hook : int
+      NM : datatype_p
+     SHM : type
+  mpi_type_free_hook : int
+      NM : datatype_p
+     SHM : type
+  mpi_op_commit_hook : int
+      NM : op_p
+     SHM : op-2
+  mpi_op_free_hook : int
+      NM : op_p
+     SHM : op-2
+  rma_op_cs_enter_hook : int
+     SHM*: win
+  rma_op_cs_exit_hook : int
+     SHM*: win
+
+PARAM:
+    addr: MPIDI_av_entry_t *
+    am_hdr: const void *
+    am_hdr_sz: size_t
+    am_hdrs: struct iovec *
+    appnum: int
+    assert: int
+    av: MPIDI_av_entry_t *
+    base: void *
+    base-2: const void *
+    base_ptr: void **
+    baseptr: void *
+    blocking: int
+    buf: const void *
+    buf-2: void *
+    buffer: void *
+    comm: MPIR_Comm *
+    comm_ptr: MPIR_Comm *
+    compare_addr: const void *
+    context_id: MPIR_Context_id_t
+    context_offset: int
+    count: MPI_Count
+    count-2: MPI_Aint
+    count-3: int
+    data: const void *
+    datatype: MPI_Datatype
+    datatype_p: MPIR_Datatype *
+    disp_unit: int *
+    disp_unit-2: int
+    displs: const int *
+    errflag: MPIR_Errflag_t *
+    flag: int *
+    group: MPIR_Group *
+    handler_id: int
+    idx: int
+    info: MPIR_Info *
+    info_p_p: MPIR_Info **
+    info_ptr: MPIR_Info *
+    init_comm: MPIR_Comm *
+    iov_len: size_t
+    is_remote: bool
+    length: MPI_Aint
+    local_upid_size: size_t **
+    local_upids: char **
+    lock_type: int
+    lpid_ptr: int *
+    lpids: const int[]
+    message: MPIR_Request *
+    message-2: MPIR_Request **
+    newcomm_ptr: MPIR_Comm **
+    newcomm_ptr-2: MPIR_Comm *
+    op: MPI_Op
+    op-2: MPIR_Op *
+    op_p: MPIR_Op *
+    origin_addr: const void *
+    origin_addr-2: void *
+    origin_count: int
+    origin_datatype: MPI_Datatype
+    partner: MPIR_Request *
+    port_name: const char *
+    port_name-2: char *
+    ptr: void *
+    rank: int
+    rdispls: const int *
+    rdispls-2: const MPI_Aint *
+    recvbuf: void *
+    recvcount: int
+    recvcounts: const int *
+    recvtype: MPI_Datatype
+    recvtypes: const MPI_Datatype[]
+    recvtypes-2: const MPI_Datatype *
+    remote_lupids: int **
+    remote_upid_size: size_t *
+    remote_upids: char *
+    req: MPIR_Request *
+    req-2: MPIR_Request **
+    request: MPIR_Request **
+    result_addr: void *
+    result_count: int
+    result_datatype: MPI_Datatype
+    root: int
+    rreq: MPIR_Request *
+    sdispls: const int *
+    sdispls-2: const MPI_Aint *
+    sendbuf: const void *
+    sendcount: int
+    sendcounts: const int *
+    sendtype: MPI_Datatype
+    sendtypes: const MPI_Datatype[]
+    sendtypes-2: const MPI_Datatype *
```

Comment:
as far as a i can tell, all instances of sendtypes-2 style arguments should be converted to `const MPI_Datatype[]`. That is the way they are specified in `mpi.h` and the MPI standard document. I haven't checked other `-2` params, but I suspect other cleanup might be possible.

Reply:
Added https://github.com/pmodels/mpich/pull/4836/commits/a0d59a7dfe137bff9cd22ffd46edf84ff8f53b21